cdap-watchdog/src/main/java/co/cask/cdap/logging/write/FileMetaDataManager.java

```diff
@@ -138,26 +138,25 @@ public int cleanMetaData(final long tillTime, final DeleteCallback callback) thr
     return execute(new TransactionExecutor.Function<Table, Integer>() {
       @Override
       public Integer apply(Table table) throws Exception {
-        byte[] tillTimeBytes = Bytes.toBytes(tillTime);
-
         int deletedColumns = 0;
         Scanner scanner = table.scan(ROW_KEY_PREFIX, ROW_KEY_PREFIX_END);
         try {
           Row row;
           while ((row = scanner.next()) != null) {
             byte[] rowKey = row.getRow();
             String namespacedLogDir = LoggingContextHelper.getNamespacedBaseDir(logBaseDir, getLogPartition(rowKey));
-            byte[] maxCol = getMaxKey(row.getColumns());
 
             for (Map.Entry<byte[], byte[]> entry : row.getColumns().entrySet()) {
               byte[] colName = entry.getKey();
               if (LOG.isDebugEnabled()) {
                 LOG.debug("Got file {} with start time {}", Bytes.toString(entry.getValue()),
                           Bytes.toLong(colName));
               }
-              // Delete if colName is less than tillTime, but don't delete the last one
-              if (Bytes.compareTo(colName, tillTimeBytes) < 0 && Bytes.compareTo(colName, maxCol) != 0) {
-                callback.handle(locationFactory.create(new URI(Bytes.toString(entry.getValue()))), namespacedLogDir);
+
+              Location fileLocation = locationFactory.create(new URI(Bytes.toString(entry.getValue())));
+              // Delete if file last modified time is less than tillTime
+              if (fileLocation.lastModified() < tillTime) {
```

Comment:
The time for the file in the meta table is for the first event in the file.
However, a log writer can continue to write log events to the same file, such that the last log line in the file is many many hours (days, even), after the first log line. So, cleanup shouldn't be based upon the timestamp of the first log line's timestamp.